Add CMake build support with Windows/MSVC compatibility

0xeb opened the pull request:
## Summary
- CMake build system with CTest integration
- Full cross-platform support (Linux, macOS, Windows)
- GitHub Actions CI for all platforms
- **Original Makefile still works on Linux**
- Cross-platform compatibility layer (`compat.h`)
- All tests pass on all platforms

## Changes
- **CMakeLists.txt**: Proper MSVC/GCC/Clang detection, CTest integration
- **compat.h**: New cross-platform header (timers, string functions, computed goto detection)
- **Source files**: Replace GCC extensions (`typeof`, computed gotos, case ranges) with portable alternatives
- **.github/workflows/ci.yml**: CI for Makefile (Linux) + CMake (Linux, macOS, Windows)

## Test plan
- [x] `make all && make test` on Linux
- [x] CMake build on Linux (GCC)
- [x] CMake build on macOS (Clang)
- [x] CMake build on Windows (MSVC)
- [x] All tests pass on all platforms
- [x] CI passing (4 jobs)

vkazanov commented:
Hi! Thanks for the suggestion!

I might merge this after checking. And if we are to extend the project then adding a comment in the README would be nice. 

But why do you think this simple demo project needs cmake at all? The Makefile is trivial.

vkazanov commented:
@0xeb , can you pls expand a bit based on my question above? Thanks

0xeb commented:
Good question and thanks for your patience. I wanted to add the CMake script as a first step to start porting this project to support Windows too.

Are you interested in supporting Windows?

If so, then CMake script will make sense as it is a universal way to prep for building, etc.

vkazanov commented:
hey, sorry for a delayed answer! December is always a busy month for me.

Either way, I don't mind support Windows per se as long as the original makefile keeps working, and instructions are provided in the README file for both build approaches.

0xeb commented:
@vkazanov it took me almost a year too. let's finish this. I did all the things you requested!

0xeb committed:
Add CMake build support with cross-platform compatibility

Add CMake build system that works on Linux, macOS, and Windows:

- CMakeLists.txt with proper compiler detection (GCC/Clang/MSVC)
- CTest integration for running tests
- GitHub Actions CI for all three platforms

Add cross-platform compatibility layer (compat.h):
- Portable timer implementation (gettimeofday vs QueryPerformanceCounter)
- POSIX to Windows function mappings (strcasecmp, strtok_r)
- COMPUTED_GOTO_SUPPORTED flag for compiler detection

Fix GCC-specific extensions for MSVC compatibility:
- Replace typeof() with memset-based initialization
- Replace case ranges (case X...Y) with default cases
- Wrap computed goto code with #if COMPUTED_GOTO_SUPPORTED
- Provide switch-based fallback for threaded interpreters on MSVC

All 7 tests pass on Linux (GCC), macOS (Clang), and Windows (MSVC).

0xeb commented:
@vkazanov now this project works across platforms. it also has CI workflow so we don't break the build. i would love if you merge.

vkazanov commented:
Hi @0xeb ,

I have a couple of review comments. Don't mind fixing these? Otherwise we're good to go.

Thanks for your work
Vlad

0xeb commented:
@vkazanov , thanks. where are your review comments? I don't see anything: 
<img width="551" height="88" alt="image" src="https://github.com/user-attachments/assets/c7abb631-8897-4c3f-afb6-ad2ab10f0bc8" />

0xeb commented:
@vkazanov I think your review comments may have been orphaned when I squashed the commits. The GitHub API shows 3 review comments exist but they're not accessible anymore.

Could you please re-add your review comments? Sorry for the inconvenience!

vkazanov commented:
Sorry, end of year adventures caught up with me. Here are the key comments:

on ci.yml:

the "cmake" branch shouldn't be mentioned there. If we are to merge this then PR-specific things should be dropped.

on the following:

  -      case 26 ... 0x1f:
  +     default:
  
 This is not right. Having an explicit value range enabled certain optimizations.

0xeb committed:
Address review feedback: fix CI branch and preserve case range optimization

- Remove 'cmake' branch from ci.yml (PR-specific, not needed after merge)
- Replace 'default:' with explicit case enumeration (case 26-31) to
  preserve compiler optimization hints while maintaining MSVC compatibility

0xeb commented:
@vkazanov Done! Fixed both issues:

1. Removed `cmake` branch from ci.yml
2. Replaced `default:` with `case 26: case 27: case 28: case 29: case 30: case 31:` — preserves optimization hints while staying portable

Let me know if this looks good to you!

0xeb committed:
Add .gitattributes to enforce consistent LF line endings

vkazanov commented:
@0xeb looks good! One last thing

One more bit: the same range is in the pigletvm-rcache.c file.

0xeb committed:
Fix case range in pigletvm-rcache.c as well

0xeb commented:
@vkazanov Good catch! Fixed pigletvm-rcache.c as well. Should be all set now.

0xeb commented:
The good thing is that now, people can easily build and use this project on most popular platforms, including Windows!

vkazanov commented:
Nice, thanks a lot! 

What got your into this little project? I haven't been working with bytecode interpreters for a while now.

0xeb commented:
Thanks for merging. 

This CTF: https://youtu.be/JfPU31HLHrc and the related videos of that challenge.
